File: F#/chap07/PlaceOrderWorkflow.fs
namespace PlaceOrderWorkflow

open Domain
open chap0702

// 注文のライフサイクル

// 検証済み
type ValidatedOrderLine = Undefined

type ValidatedOrder =
    { OrderId: OrderId
      CustomerInfo: CustomerInfo
      ShippingAddress: Address
      BillingAddress: Address
      OrderLines: ValidatedOrderLine list }

and OrderId = Undefined
and CustomerInfo = Undefined
and Address = Undefined

// 計算済み
type PricedOrderLine = Undefined
type PricedOrder = Undefined

// 注文の全状態
type Order =
    | Unvalidated of UnvalidatedOrder
    | Validated of ValidatedOrder
    | Priced of PricedOrder

// 内部ステップ

// 注文の検証
type ProductCode = Undefined
type CheckProductCodeExists = ProductCode -> bool

type AddressValidationError = Undefined
type CheckedAddress = Undefined
type CheckAddressExists = UnvalidatedAddress -> AsyncResult<CheckedAddress, AddressValidationError>

type ValidateOrder =
    CheckProductCodeExists // 依存関係
        -> CheckAddressExists // 依存関係
        -> UnvalidatedOrder // 入力
        -> AsyncResult<ValidatedOrder, ValidationError list>

and ValidationError = Undefined

// 注文の価格計算
type Price = Undefined
type GetProductPrice = ProductCode -> Price
type PricingError = Undefined

type PriceOrder =
    GetProductPrice // 依存関係
        -> ValidatedOrder // 入力
        -> Result<PricedOrder, PricingError>

File: Scala/chap07/src/main/scala/PlaceOrderWorkflow.scala
import Domain.{UnvalidatedAddress, UnvalidatedOrder}
import cats.data.EitherT

import scala.concurrent.Future

object PlaceOrderWorkflow:

  // 検証済みデータ
  case class ValidatedOrderLine(value: Any)

  case class OrderId(value: Any)

  case class CustomerInfo(value: Any)

  case class Address(value: Any)

  case class ValidatedOrder(
                             orderId: OrderId,
                             customerInfo: CustomerInfo,
                             shippingAddress: Address,
                             billingAddress: Address,
                             orderLines: List[ValidatedOrderLine],
                           )

  // 計算済みデータ
  case class PricedOrderLine(value: Any)

  case class PricedOrder(value: Any)

  // 注文の全状態
  enum Order:
    case Unvalidated(value: UnvalidatedOrder)
    case Validated(value: ValidatedOrder)
    case Priced(value: PricedOrder)

  // 注文の検証
  case class ProductCode(value: Any)

  type CheckProductCodeExists = ProductCode => Boolean

  case class AddressValidationError(value: Any)

  case class CheckedAddress(value: Any)

  type CheckAddressExists = UnvalidatedAddress => EitherT[Future, AddressValidationError, CheckedAddress]

  case class ValidationError(value: Any)

  type ValidateOrder =
    CheckProductCodeExists // 依存
      => CheckAddressExists // 依存
      => UnvalidatedOrder // In
      => EitherT[Future, List[ValidationError], ValidatedOrder] // Out

  // 価格計算
  case class Price(value: Any)

  type GetProductPrice = ProductCode => Price

  case class PricingError(value: Any)

  type PriceOrder =
    GetProductPrice // 依存
      => ValidateOrder // In
      => Either[PricingError, PricedOrder] // Out
